### ISSUE TITLE
Pagination support in dbcore

### ISSUE THREAD

**Issue by PierreRust** [beetbox/beets#750]
The creation of this issue has been discussed in #736.
To implement pagination properly in the HTTP API, we need to add support for pagination in dbCore. Ideally the pagination mechanism should also be usable with the command-line interface.

Ideas about pagination have been discussed in #718. It seems were are basically 4 options : 
- the naive approach : using start/offset parameters in query. While this is trivial to implement it is definitely not very efficient.
- use a seek-able id in the query ( i.e. "gives me the next items starting with this one") : this is the approach [recommended by sqlite](http://www.sqlite.org/cvstrac/wiki?p=ScrollingCursor), we have to make a small prototype to make sure it's doable with complex queries.
- when performing a query, store temporarily the result and return a token to the client, which can be used to walk through the ResultSet. The difficult part here is to maintain, and discard at some point, the ResultSets
- last solution : only return id for complex queries. This looks nice but I'm afraid it might not play well with the current query mechanism, which fetches full rows anyway.

**Comment by pjones** [comment]
This would be especially useful for the Smart Playlist and Random plug-ins.  Right now there's no way to limit the number of tracks generated by the Smart Playlist plug-in.

**Comment by jonthomas83** [comment]
Sorry guys, I'm not sure if I'm barking up the wrong tree here, and please tell me to get lost, but from a UI perspective, pagination or pager navigation isn't the most user friendly interface to have to use - we've done a lot of usability testing and found that they're not as straightforward as we'd like to think. You'd be surprised at the types of feedback, which I can share if you need me to.

A lot of the stuff mentioned here goes right over my head, but after reading https://github.com/sampsyo/beets/issues/738, it seems like pagers/paged navigation methods will play a part in the proposed UI implementations. Can we reconsider their use and look at alternative options before making a decision - like I said, in terms of UI. Many thanks.

**Comment by sampsyo** [comment]
Yes! I absolutely agree that it's not a foregone conclusion that the interface should paginate. But it's still useful for the API to admit pagination for efficiency reasons—independent of the interface.

**Comment by jonthomas83** [comment]
That's good news, as long as we're on the same "page" ;-)

FWIW, our guys are still developing paging on the API too, makes sense.

**Comment by semohr** [comment]
I’m not sure if this issue is still relevant, but I just came across it.
I recently implemented a paginated query that might be reusable for this purpose, see [here](https://github.com/pSpitzner/beets-flask/blob/792a4f9c61eaa8293d481416228205b8bc0fc3dd/backend/beets_flask/server/routes/library/resources.py?utm_source=chatgpt.com#L481-L541).
